core/scratch_block_comment.js

```diff
@@ -0,0 +1,422 @@
+
+goog.require('goog.math.Coordinate');
+goog.require('goog.userAgent');
+
+
+/**
+ * Class for a comment.
+ * @param {!Blockly.Block} block The block associated with this comment.
+ * @param {number=} x Initial x position for comment, in workspace coordinates.
+ * @param {number=} y Initial y position for comment, in workspace coordinates.
+ * @param {boolean=} minimized Whether or not this comment is minimized
+ *     (only the top bar displays), defaults to false.
+ * @extends {Blockly.Comment}
+ * @constructor
+ */
+Blockly.ScratchBlockComment = function(block, x, y, minimized) {
+  Blockly.ScratchBlockComment.superClass_.constructor.call(this, block);
+  this.x_ = x;
+  this.y_ = y;
+  this.isMinimized_ = minimized || false;
+};
+goog.inherits(Blockly.ScratchBlockComment, Blockly.Comment);
+
+/**
+ * Width of bubble.
+ * @private
+ */
+Blockly.ScratchBlockComment.prototype.width_ = 200;
+
+/**
+ * Height of bubble.
+ * @private
+ */
+Blockly.ScratchBlockComment.prototype.height_ = 200;
+
+/**
+ * Comment Icon Size.
+ */
+Blockly.ScratchBlockComment.prototype.SIZE = 0;
+
+/**
+ * Offset for text area in comment bubble.
+ * @private
+ */
+Blockly.ScratchBlockComment.TEXTAREA_OFFSET = 12;
+
+/**
+ * Maximum lable length (actual label length will include
+ * one additional character, the ellipsis).
+ * @private
+ */
+Blockly.ScratchBlockComment.MAX_LABEL_LENGTH = 16;
+
+/**
+ * Width that a minimized comment should have.
+ * @private
+ */
+Blockly.ScratchBlockComment.MINIMIZE_WIDTH = 200;
+
+/**
+ * Draw the comment icon.
+ * @param {!Element} _group The icon group.
+ * @private
+ */
+Blockly.ScratchBlockComment.prototype.drawIcon_ = function(_group) {
+  // NO-OP -- Don't render a comment icon for Scratch block comments
+};
+
+// Override renderIcon from Blocky.Icon so that the comment bubble is
+// anchored correctly on the block. This function takes in the top margin
+// as an input instead of setting an arbitrary one.
+/**
+ * Render the icon.
+ * @param {number} cursorX Horizontal offset at which to position the icon.
+ * @param {number} topMargin Vertical offset from the top of the block to position the icon.
+ * @return {number} Horizontal offset for next item to draw.
+ * @package
+ */
+Blockly.ScratchBlockComment.prototype.renderIcon = function(cursorX, topMargin) {
+  if (this.collapseHidden && this.block_.isCollapsed()) {
+    this.iconGroup_.setAttribute('display', 'none');
+    return cursorX;
+  }
+  this.iconGroup_.setAttribute('display', 'block');
+
+  var width = this.SIZE;
+  if (this.block_.RTL) {
+    cursorX -= width;
+  }
+  this.iconGroup_.setAttribute('transform',
+      'translate(' + cursorX + ',' + topMargin + ')');
+  this.computeIconLocation();
+  if (this.block_.RTL) {
+    cursorX -= Blockly.BlockSvg.SEP_SPACE_X;
+  } else {
+    cursorX += width + Blockly.BlockSvg.SEP_SPACE_X;
+  }
+  return cursorX;
+};
+
+/**
+ * @typedef CommentContent The comment editor and label text to display in the
```

Comment:
Does this typedef actually work?  It's different from this way of doing a typedef: https://github.com/google/closure-compiler/wiki/Using-JSDoc-Annotations-with-Closure-compiler#typedef